core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -378,55 +381,157 @@ public Object visitGetAllCommand(InvocationContext ctx, GetAllCommand command) t
       }
 
       LocalizedCacheTopology cacheTopology = checkTopologyId(command);
-      Map<Address, List<Object>> requestedKeys = getKeysByOwner(ctx, command.getKeys(), cacheTopology, null);
+      Map<Address, List<Object>> requestedKeys = getKeysByOwner(ctx, command.getKeys(), cacheTopology, null, null);
       if (requestedKeys.isEmpty()) {
          return invokeNext(ctx, command);
       }
 
-      CompletableFuture<Void> allFuture = remoteGetAll(ctx, command, requestedKeys);
-      return asyncInvokeNext(ctx, command, allFuture);
-   }
-
-   protected CompletableFuture<Void> remoteGetAll(InvocationContext ctx, GetAllCommand command, Map<Address, List<Object>> requestedKeys) {
       GlobalTransaction gtx = ctx.isInTxScope() ? ((TxInvocationContext) ctx).getGlobalTransaction() : null;
-      CompletableFutureWithCounter allFuture = new CompletableFutureWithCounter(requestedKeys.size());
+      ClusteredGetAllFuture allFuture = new ClusteredGetAllFuture(requestedKeys.size(), command);
 
       for (Map.Entry<Address, List<Object>> pair : requestedKeys.entrySet()) {
-         List<Object> keys = pair.getValue();
-         ClusteredGetAllCommand clusteredGetAllCommand = cf.buildClusteredGetAllCommand(keys, command.getFlagsBitSet(), gtx);
+         ClusteredGetAllCommand clusteredGetAllCommand = cf.buildClusteredGetAllCommand(pair.getValue(), command.getFlagsBitSet(), gtx);
          clusteredGetAllCommand.setTopologyId(command.getTopologyId());
-         rpcManager.invokeRemotelyAsync(Collections.singleton(pair.getKey()), clusteredGetAllCommand, syncIgnoreLeavers).whenComplete((responseMap, throwable) -> {
-            if (throwable != null) {
-               allFuture.completeExceptionally(throwable);
-            }
-            if (responseMap.isEmpty()) {
-               allFuture.completeExceptionally(OutdatedTopologyException.INSTANCE);
+         Set<Address> target = Collections.singleton(pair.getKey());
+         rpcManager.invokeRemotelyAsync(target, clusteredGetAllCommand, syncIgnoreLeavers)
+               .whenComplete(new ClusteredGetAllHandler(allFuture, ctx, command, pair.getValue(), target));
+      }
+      return asyncValue(handleRemotelyRetrievedKeys(ctx, allFuture, allFuture)).thenApply(ctx, command, allFuture);
```

Comment:
actually this is wrong; we need to handle remotely retrieved keys only *after* we retrieve them